### PR TITLE
importccl: Resume CSV imports

### PR DESCRIPTION
**Pull request by miretskiy**
Add ability to resume csv imports based on the state persisted in the
job progress record.

Add an end-to-end test verifying this functionality.

Release note (cli change): CSV import can be resumed from checkpointed state.

### CODE REVIEW

**Review by dt — COMMENTED** [review]
Nice stuff! Some of these test mocks look really useful in the broader context of testing jobs, either as reusable pieces or at least as reference patterns. After this merges, maybe we can sit down with Spas and identify a suite of tools / patterns that make sense when testing the behavior of stateful jobs.

I didn't fully review the rest of the test since I did have some nits about some of the changes / knobs to the non-test code to facilitate those tests, but overall this looks pretty good!

> Reviewed 1 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-:-Lu5xLUy1n53dTCvnjAZ:bf0rhw8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/importccl/import_processor.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5wNuxC-EyCo4kzrMK:-Lu5wNuxC-EyCo4kzrML:b-b6vz8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/ccl/importccl/import_processor.go#L340)):*
> ```Go
> 			writtenRow[offset] = kvBatch.LastRow
> 			atomic.StoreUint32(&writtenFraction[offset], math.Float32bits(kvBatch.Progress))
> 			if flowCtx.Cfg.TestingKnobs.BulkAdderFlushesEveryBatch {
> ```

I'm not a huge fan of this -- this means we're _not_ testing that the actual, normal threshold-based progress sending is doing its job.

I'd be more okay with a more synthetic case for the sake of testing if the goal weren't to be end-to-end, but in that case I think we could just use unit-tests of the individual pieces instead as well. Since the reason we're doing all this is an end-to-end test, I'd prefer to try to have the code run as similarly as we can to production.

Could we try plumbing a buffer size limit of 1 byte to trigger flushes that way instead?
___
*[pkg/ccl/importccl/testutils_test.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lte5ExwDNUn4rYfeIWe:-Lte5ExwDNUn4rYfeIWf:b3tu8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5c0e10cb3fd7891fdffa416f19c32c1373f53b/pkg/ccl/importccl/testutils_test.go#L109)):*
> ```Go
> 	// An error result will cause csv generation to stop and the csv server to return that error.
> 	csvBpHandler  = func() (bool, error)
> 	csvBreakpoint struct {
> ```

2¢: while the large type block is technically allowed, it is unusual. I think I prefer seeing `type ... struct {` each time.
___
*[pkg/ccl/importccl/testutils_test.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lte4w3w6JKUCsvAGBbg:-Lte4w3w6JKUCsvAGBbh:bss78ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5c0e10cb3fd7891fdffa416f19c32c1373f53b/pkg/ccl/importccl/testutils_test.go#L139)):*
> ```Go
> }
> 
> // Note: we read one row at a time because reading as much as the buffer space allows might cause problems
> ```

here / elsewhere: style guide says comments wrapped at 80.
___
*[pkg/ccl/importccl/testutils_test.go, line 284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu546KACbTueUKQ-5dh:-Lu546KACbTueUKQ-5di:b-g7vnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/ccl/importccl/testutils_test.go#L284)):*
> ```Go
> 
> // Returns the list of file names (URIs) suitable for this factory.
> func (ses *generatedStorage) getGeneratorURIs() []interface{} {
> ```

my kingdom for covariant containers
___
*[pkg/sql/distsql_plan_csv.go, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vDK-9lnBv14jMTz2:-Lu5vDK-9lnBv14jMTz3:b-z8pvj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L216)):*
> ```Go
> 	format roachpb.IOFileFormat,
> 	walltime int64,
> 	alwaysFlushProgress bool,
> ```

Having to add this to the public, production, runtime API is unfortunate. 

I'd love to see if we could find another approach (also considering nit below).
___
*[pkg/sql/distsql_plan_csv.go, line 296 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vFZj8tV9Gb-HP_6C:-Lu5vFZj8tV9Gb-HP_6D:b9dxpbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L296)):*
> ```Go
> 
> 			if alwaysFlushProgress {
> 				_ = updateJobProgress()
> ```

hm, not a big fan of swallowing this error, or doing this work on the receiver goroutine.

**Review by spaskob — COMMENTED** [review]
I can only see 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-:-Lu6aXUM4YennvZ20ZSK:b-3o32wb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @spaskob)

___
*[pkg/sql/distsql_plan_csv.go, line 296 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vFZj8tV9Gb-HP_6C:-Lu6axIY3lU3olevWAZt:b-f55k2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L296)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

hm, not a big fan of swallowing this error, or doing this work on the receiver goroutine.
</blockquote></details>

Is this a glitch in Reviewable?  I am not seeing the second commit.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/sql/distsql_plan_csv.go, line 296 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vFZj8tV9Gb-HP_6C:-Lu6d961EqI64PEPNT7i:bfeckpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L296)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Is this a glitch in Reviewable?  I am not seeing the second commit.
</blockquote></details>

Nevermind, reviewable was set to review each commit separately

**Review by miretskiy — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, 7 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/importccl/import_processor.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5wNuxC-EyCo4kzrMK:-Lu9qzuT2oZsm9I3VZvF:b-i5n4l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/ccl/importccl/import_processor.go#L340)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'm not a huge fan of this -- this means we're _not_ testing that the actual, normal threshold-based progress sending is doing its job.

I'd be more okay with a more synthetic case for the sake of testing if the goal weren't to be end-to-end, but in that case I think we could just use unit-tests of the individual pieces instead as well. Since the reason we're doing all this is an end-to-end test, I'd prefer to try to have the code run as similarly as we can to production.

Could we try plumbing a buffer size limit of 1 byte to trigger flushes that way instead?
</blockquote></details>

This is a tough one, David.

First, allow me to say that I'm not a huge fan of this either, or of the testing knobs in general (see my attempted "testvalue" proposal pr).  

I have attempted to do just  as you suggested, via changing various KV settings (via set cluster settings), basically unsuccessfully.   
Setting batch sizes to 1 will certainly cause flushes to happen (and they did).  However, those flushes would pick up "stale" value for the row -- because we update written row after we finish with the batch.  This made tests pretty hard to follow (i.e. I flush at position X, but I expect progress to be X minus delta).

In addition, and more importantly, this knob doesn't actually short circuit batching behavior (if anything -- uses more "normal" code path by not futzing with low level sst flushing).  This knob controls a "timer" that causes progress to be pushed.  It's this timer that was SO problematic when trying to get these tests to be deterministic.  Changing sleep parameter to any value, no matter how low, produced non-deterministic tests, particularly since I"m generating data on the fly.  In addition, it was *very* hard to synchronize this CSV generator with non-deterministic timer.  Scratch that: I was never able to get it to work deterministically.

Since changing sleep parameter is not an option, perhaps adding a testing knob to control flush timer might make sense.  Alas, I adding this kind of a knob is even worse (who controls the timer? who stops it?.. it was a lot worse than just a boolean parameter).

Another possibility that I (admittedly) briefly entertained was refactoring the "progress push" go routine.  This refactoring would still be pretty hard to hook up because, ultimately, if I want to have an "IMPORT INTO..." type test, I would have to specify this injection somehow -- probably with another testing knob.

To summarize: I agree with you; this thing sucks.  I have tried various approaches to work my way around, and, on the balance, I believe this approach is the lesser evil of the other ugly options.

Please let me know if you see something that I don't -- I'd be willing to give it another go.

   
___
*[pkg/ccl/importccl/testutils_test.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lte5ExwDNUn4rYfeIWe:-Lu9mEw18_iQqhQOVDed:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5c0e10cb3fd7891fdffa416f19c32c1373f53b/pkg/ccl/importccl/testutils_test.go#L109)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

2¢: while the large type block is technically allowed, it is unusual. I think I prefer seeing `type ... struct {` each time.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/testutils_test.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lte4w3w6JKUCsvAGBbg:-Lu9os_s09GwK_xKf8iJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5c0e10cb3fd7891fdffa416f19c32c1373f53b/pkg/ccl/importccl/testutils_test.go#L139)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

here / elsewhere: style guide says comments wrapped at 80.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/testutils_test.go, line 284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu546KACbTueUKQ-5dh:-Lu9pCvG1pPjRSctTqj9:b17hlje) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/ccl/importccl/testutils_test.go#L284)):*
>  covariant containers
Yeah... Just makes it easy to use at the call site; no need to convert.
___
*[pkg/sql/distsql_plan_csv.go, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vDK-9lnBv14jMTz2:-LuAk3hU-bVH0N4l9NFJ:b-dv8oz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L216)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Having to add this to the public, production, runtime API is unfortunate. 

I'd love to see if we could find another approach (also considering nit below).
</blockquote></details>

Agreed, see my comment re bulk flushing testing knob and let's see what we can do here.
___
*[pkg/sql/distsql_plan_csv.go, line 296 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-Lu5vFZj8tV9Gb-HP_6C:-LuAHgL28yaO3XWjQy7w:bekfh9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/aea276ce4de45479c296a978d5d79e445e738e02/pkg/sql/distsql_plan_csv.go#L296)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Nevermind, reviewable was set to review each commit separately
</blockquote></details>

Updated metaFn definition(s) to make it possible to return an error.

**Review by miretskiy — COMMENTED** [review]
> Reviewed 1 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

**Comment by miretskiy** [comment]
@dt Mind taking another look?

**Comment by miretskiy** [comment]
rebase onto master...

**Review by dt — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/importccl/read_import_csv.go, line 107 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-LvfKcuz4_6FmqK9fAiG:-LvfKcuz4_6FmqK9fAiH:b-8oa6zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e2e5a25b605c328e9181ded6c18bd3afa7ad0bb/pkg/ccl/importccl/read_import_csv.go#L107)):*
> ```Go
> 		ctx, span := tracing.ChildSpan(ctx, "convertcsv")
> 		defer tracing.FinishSpan(span)
> 		return ctxgroup.GroupWorkers(
> ```

This has bitten me before, and I think we should just change the signature of the worker func to get an ID passed by `GroupWorkers` rather than introduce this mutex-protected counter.

**Review by miretskiy — COMMENTED** [review]
> Reviewed 6 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42476)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/importccl/read_import_csv.go, line 107 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42476#-LvfKcuz4_6FmqK9fAiG:-Lvg49mMBifNk-d5AKls:b-ucpoxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e2e5a25b605c328e9181ded6c18bd3afa7ad0bb/pkg/ccl/importccl/read_import_csv.go#L107)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This has bitten me before, and I think we should just change the signature of the worker func to get an ID passed by `GroupWorkers` rather than introduce this mutex-protected counter.
</blockquote></details>

Done (in a separate commit)

**Comment by jseldess** [comment]
@miretskiy, based on the code, I can't figure out how this works from a user-facing perspective. CAn you please provide more details? Is there a new `IMPORT` option or something?

**Comment by miretskiy** [comment]
Hi Jesse,

No problem, glad to be of help.  There are no new import options.  Instead,
the import job
(whether it's CSV or any other supported format) periodically updates its
state in the system.jobs table.  This state, among other things, includes
information indicating how much of the input data was processed.

If the job is restarted, we pick up import from that checkpoint position in
the input.
